### ISSUE TITLE
Too many users can edit discussion responses

### ISSUE THREAD

**Issue by mrclay** [Elgg/Elgg#9762]
It was [discovered](https://elgg.org/discussion/view/2485285/is-it-a-bug-or-a-questionable-feature-that-users-can-edit-all-the-comments-in-discussions-they-create) that group owners and discussion topic owners can edit replies. This seems to be new behavior added in 1.9, in a [commit for an unrelated feature](https://github.com/Elgg/Elgg/commit/6e31220b882a7698abbb35356504fede8fcbc4b3#diff-d69a9c513b5e301164eb156de0de8db5R1067) (ajax edit).

This is probably not a sensible default, and there doesn't seem to exist an obvious default behavior, so I think setting(s) should be added to define it.

**Comment by hypeJunction** [comment]
I think we had similar issues with comments, so it would make sense starting there.

**Comment by mrclay** [comment]
Strawman settings:

In core:
A. [ ] Users can edit any comment on their content.
B. [ ] Users can delete any comment on their content.

In groups:
C. [ ] Group owners can edit any comment/discussion reply within the group
D. [ ] Group owners can delete any comment/discussion reply within the group

In discussions:
E. [ ] Users can edit any reply to their discussion topic.
F. [ ] Users can delete any reply to their discussion topic.

In likes:
G: [ ] Users can delete likes on their content.

**Comment by iionly** [comment]
This has just been asked about on the community site (https://elgg.org/discussion/view/2596157/problem-with-plugin-discussion).

I think this is a serious issue because if any user can modify comments / discussions replies of other users it is likely to cause conflicts on Elgg communities if only one or a few users utilize these possibilites either just to annoy other users or to get some control within a discussion (maybe even by changing the postings of other users and then complain about being attacked in an unfair manner).

In my opionion:

- Core: only admins and not normal users should be able to edit or delete comments of other users. A user should NEVER be able to edit comment of other users and I think deletion should also not be possible. If a normal user feels a comment is wrong there's always the option to report it to an admin who should deal with it then.
- Groups: the group owners are kind of admins within their groups. So I think they should have permission to edit and delete content of other users within their groups in addition to site admins. But again: surely no normal group member should be able to edit or delete comments of other users or discussion replies.
- Discussions: with the discussions only available within groups the permissions for editing and deleting discussions replies would have been admin and group owner respectively. With discussions also available separately, the edit/delete permissions would have to be restricted to site admins for the separate discussions.
- Likes: if a user wants to delete likes made by others on their own content. Well, let them.

**Comment by hypeJunction** [comment]
I agree with the above. We should put this logic into separate hooks so it's easy to unregister them, and we don't really need settings then.

**Comment by hypeJunction** [comment]
Do you want to tackle it, @iionly?

**Comment by iionly** [comment]
I think I'll give it a try. No promise that I can handle it though. I might split it into separate PRs for comments, discussions and likes.

Though it seems after testing briefly that comments might already behave as expected.

Lkes deletion seems to me a feature PR.

Discussion replies edit/delete permissions seems to me rather a bugfix. If I'm not mistaken the edit/delete permissions for replies by the topic starter was added in https://github.com/Elgg/Elgg/commit/6e31220b882a7698abbb35356504fede8fcbc4b3#diff-d69a9c513b5e301164eb156de0de8db5 (and very likely by mistake as @juho-jaakkola included some unused callback function stub from 1.8 into 1.9).

Are there any problems to be expected if the disussion topic starter wants to delete the discussion topic if he no longer has edit permission on the replies (as delete permissions would be the same when no separate hook handler is defined)? I think I got the edit/delete permissions working already as expected simply by commenting out 3 lines in the already existing plugin hook handler but I wonder about what happens with the discussions replies when the discussion owner deletes the discussion entity because there's nothing done with the replies in the discussion delete action. Are they automatically deleted when the container entity is deleted regardless of different owner?

**Comment by hypeJunction** [comment]
Just go with your gut feeling. We can always give feedback if we feel that there are other use cases. At this point let's put everything into master, no reason to complicate things with BC, break it if you have to.

**Comment by iionly** [comment]
I've made 3 PRs to update edit/delete permissions for comments, discussion_replies and likes respectively: https://github.com/Elgg/Elgg/pull/10702, https://github.com/Elgg/Elgg/pull/10703 and https://github.com/Elgg/Elgg/pull/10704.

In all 3 cases an appropriate plugin hook callback function was already available. So, I just updated these. If a site owner wants different edit/delete permissions, he can do so by simple replacing them.

**Comment by iionly** [comment]
I've decided to close my PRs that would have modified the edit/delete permissions for comments, discussion replies and likes respectively. I'll make a plugin for that instead.
